---
title: "QTL mapping with R/qtl"
teaching: 0
exercises: 0
questions:
- "?"
objectives:
- 
- 
keypoints:
- "."
- "."
source: Rmd
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("02-")
```

Hypertension, or high blood pressure, affects more than one-fourth of the U.S. population and is a leading cause of heart attack and stroke. Hypertension is complex, is influenced by many genes, and is exacerbated by elevated sodium intake among other lifestyle factors. [Sugiyama et al. 2001](https://s3.amazonaws.com/academia.edu.documents/45963759/geno.2000.640120160526-29022-36mpgg.pdf?AWSAccessKeyId=AKIAIWOWYYGZ2Y53UL3A&Expires=1502397248&Signature=hVoWIHwrYd7KI0UNIaSm7kWviWk%3D&response-content-disposition=inline%3B%20filename%3DConcordance_of_murine_quantitative_trait.pdf) investigated the genetics of salt-induced hypertension. 250 male mice from a reciprocal backcross between salt-sensitive C57BL/6J and resistant A/J strains were treated with a 1% salt solution in drinking water. QTL analysis identified 6 loci contributing to variation in hypertension. 

Data for this study are included in the `qtl` package. To access the data and perform QTL analysis, first load the `qtl` library, then use the `data` function to load the built-in dataset. Once the data are loaded, we'll begin exploration of the data and the capabilities of the `qtl` package.

```{r}
library(qtl)
data(hyper)
```

Use `ls` to list the components of the data set now that they are loaded into your environment. Alternatively, you can view them in the `Environment` panel in RStudio. `?hyper` will bring you to the documentation describing these data.

```{r}
ls()
?hyper
```

The `summary` function prints summary information about the data set. 

```{r}
summary(hyper)
```

Several other functions provide specific information about the data, such as the number of phenotypes, individuals, chromosomes, and markers. 

```{r}
nind(hyper)
nphe(hyper)
nchr(hyper)
totmar(hyper)
nmar(hyper)
```

Plot a summary of these data.

```{r}
plot(hyper)
```

Let's look at each panel individually, starting with missing genotypes.

```{r}
plotMissing(hyper)
```

In the upper left, black pixels indicate missing genotype data. Can you find the one marker that has no genotype data? Notice that fewer than 100 individuals were genotyped for most of the markers. 

Now let's view a genetic map of the markers. 

```{r}
plotMap(hyper)
```

Notice that on some chromosomes, such as 18 and 19, there are very few markers. Others, like chromosome 4, have many more markers. This study was published in 2001 when genotyping many markers was relatively expensive. At present we expect much greater marker density and far more genotyping. Nevertheless a sparse genetic map like this serves to teach concepts that we need to understand for denser genotyping data.

Plot the missing data again, re-ordering the individuals according to the value of the blood pressure phenotype.

```{r}
plotMissing(hyper, reorder=TRUE)
```

For most markers, only individuals with extreme phenotypes were genotyped. At many markers (in regions of interest), markers were typed only on recombinant individuals.

We can also view a histogram of the blood pressure phenotype, as well as a chart indicating the number of animals of each sex.

```{r}
plotPheno(hyper, pheno.col = 1)
plotPheno(hyper, pheno.col = 2)
```

We now, finally, get to QTL mapping.The core of R/qtl is a set of functions which make use hidden Markov models to calculate QTL genotype probabilities, to simulate from the joint genotype distribution and to calculate the most likely sequence of underlying genotypes (all conditional on the observed mark
er data). This is done in a quite general way, with possible allowance for the presence of genotyping errors. Of course, for convenience we assume no crossover interference.

The function `calc.genoprob` calculates QTL genotype probabilities, conditional on the available marker data. These are needed for most of the QTL mapping functions. The argument `step` indicates the step size (in cM) at which the probabilities are calculated, and determines the step size at which later LOD scores are calculated.

```{r}
hyper <- calc.genoprob(hyper, step=1, error.prob=0.01)
```

We may now use the function `scanone` to perform a single-QTL genome scan with a normal model. We may use maximum likelihood via the EM algorithm (Lander and Botstein 1989) or use Haley-Knott regression (Haley and Knott 1992).

```{r}
out.em <- scanone(hyper)
out.hk <- scanone(hyper, method="hk")
```

The output of scanone has class `scanone`; the function `summary.scanone` displays the maximum LOD score on each chromosome for which the LOD exceeds a specified threshold.

```{r}
summary(out.em)
summary(out.em, threshold=3)
summary(out.hk, threshold=3)
```

The function `max.scanone` returns just the highest peak from output of `scanone`.

```{r}
max(out.em)
max(out.hk)
```

We may also plot the results. `plot.scanone`
can plot up to three genome scans at once, provided that they conform appropriately. Alternatively, one may use the argument `add`.

```{r}
plot(out.em, chr=c(1,4,15))
plot(out.em, out.hk, chr=c(1,4,15))
plot(out.em, chr=c(1,4,15))
plot(out.hk, chr=c(1,4,15), col="blue", add=TRUE)
```

The function `scanone` may also be used to perform a permutation test to get a genome-wide LOD significance threshold. For Haley-Knott regression, this can be quite fast.

```{r}
operm.hk <- scanone(hyper, method="hk", n.perm=1000)
```

The permutation output has class `scanoneperm`. The function `summary.scanoneperm` can be used to get
significance thresholds.

```{r}
summary(operm.hk, alpha=0.05)
```

In addition, if the permutations results are included in a call to `summary.scanone`, you can estimate genome-scan-adjusted p-values for inferred QTL, and can get a report of all chromosomes meeting a certain significance level, with the corresponding LOD threshold calculated automatically.

```{r}
summary(out.hk, perms=operm.hk, alpha=0.05, pvalues=TRUE)
```

The function `scantwo` performs a two-dimensional genome scan with a two-QTL model. For every pair of positions, it calculates a LOD score for the full model (two QTL plus interaction) and a LOD score for the additive model (two QTL but no interaction). This can be quite time consuming, and so you may wish to do the calculations on a coarser grid.

```{r, results='hide'}
hyper <- calc.genoprob(hyper, step=5, error.prob=0.01)
out2.hk <- scantwo(hyper, method="hk")
```

One can also use method="em" or method="imp", but they are even more time consuming.

The output of `scantwo` has class `scantwo`; there are functions for obtaining summaries and plots, of course.

The summary function considers each pair of chromosomes, and calculates the maximum LOD score for the full model (Mf) and the maximum LOD score for the additive model (Ma). These two models are allowed to be maximized at different positions. We further calculate a LOD score for a test of epistasis, Mi=Mf−Ma, and two LOD scores that concern evidence for a second QTL: Mfv1 is the LOD score comparing the full model to the best single-QTL model and Mav1
is the LOD score comparing the additive model to the best single-QTL model.

In the summary, we must provide five thresholds, for Mf, Mfv1, Mi, Ma, and Mav1, respectively. Call these Tf, Tfv1, Ti, Ta, and Tav1. We then report those pairs of chromosomes for which at least
one of the following holds:
7•Mf≥Tf and (Mfv1≥Tfv1 or Mi≥Ti)•Ma≥Ta and Mav1≥Tav1
The thresholds can be obtained by a permutation test (see below), but this is extremely time-consuming. For a mouse backcross, we suggest the thresholds (6.0, 4.7, 4.4, 4.7, 2.6) for the full, conditional-interactive, interaction, additive, and
conditional-additive LOD scores, respectively. For a mouse intercross, we suggest the thresholds (9.1, 7.1, 6.3, 6.3, 3.3) for the full, conditional-interactive, interaction, additive, and conditional-additive LOD scores, respectively. These were obtained by 10,000 simulations of crosses with 250 individuals, markers at a 10 cM spacing, and analysis by Haley-Knott regression.

```{r}
summary(out2.hk, thresholds=c(6.0, 4.7, 4.4, 4.7, 2.6))
```

The appropriate decision rule is not yet completely clear.  I am inclined to ignore Mi and to choose genome-wide thresholds for the other four based on a permutation, using a common significance level for all four. Mi would be ignored if we gave it a very large threshold, as follows.

```{r}
summary(out2.hk, thresholds=c(6.0, 4.7, Inf, 4.7, 2.6))
```

Plots of `scantwo` results are created via `plot.scantwo`.

```{r}
plot(out2.hk)
plot(out2.hk, chr=c(1,4,6,15))
```

By default, the upper-left triangle contains epistasis LOD scores and the lower-right triangle contains the LOD scores for the full model. The color scale on the right indicates separate scales for the epistasis and joint LOD scores (on the left and right, respectively).

The function `max.scantwo` returns the two-locus positions with the maximum LOD score for the full and additive models.

```{r}
max(out2.hk)
```

One may also use `scantwo` to perform permutation tests in order to obtain genome-wide LOD significance thresholds. These can be extremely time consuming, though with the Haley-Knott regression and multiple imputation methods, there is a trick that may be used in some cases to dramatically speed things up. So we’ll try 100 permutations by the Haley-Knott regression method and hope that your computer is sufficiently fast.

```{r}
operm2.hk <- scantwo(hyper, method="hk", n.perm=100)
```

We can again use `summary` to get LOD thresholds.

```{r}
summary(operm2.hk)
```

And again these may be used in the summary of the
`scantwo` output to calculate thresholds and p-values. If you want to ignore the LOD score for the interaction in the rule about what chromosome pairs to report, give α = 0, corresponding to a threshold T=∞.

```{r}
summary(out2.hk, perms=operm2.hk, pvalues=TRUE,
alphas=c(0.05, 0.05, 0, 0.05, 0.05))
```

You can’t really trust these results. Haley-Knott regression performs poorly in the case of selective genotyping (as with the hyper data). Standard interval mapping or imputation would be better, but Haley-Knott regression has the advantage
of speed, which is the reason we use it here.
