---
title: "Calculating Genotype Probabilities"
teaching: 0
exercises: 0
questions:
- "How do I calculate QTL at positions between genotyped markers?"
- "How do I calculate QTL genotype probabilities?"
- "How do I calculate allele probabilities?"
objectives:
- To understand that the first step in QTL analysis is to calculate genotype probabilities.
- To insert pseudomarkers between genotyped markers.
- To calculate genotype probabilities.
- To specify the number of cores to use in order to speed up calculations.
keypoints:
- "The first step in QTL analysis is to calculate genotype probabilities."
- "Insert pseudomarkers to calculate QTL at positions between genotyped markers."
- "Calculate genotype or allele probabilities between genotyped markers with calc_genoprob() or genoprob_to_alleleprob()."
source: Rmd
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("06-")
```

The first basic task in QTL analysis is to calculate conditional genotype probabilities, given the observed marker data, at each putative QTL position. This is accomplished with the `calc_genoprob()`
function in the [qtl2geno](https://github.com/rqtl/qtl2geno)
package. Unlike the corresponding function in
[R/qtl](http://rqtl.org), `calc.genoprob()`, the result is not inserted back into the input cross  object, but is returned as a list of three-dimensional
arrays (one per chromosome). Each 3d array of probabilities is arranged as individuals &times; genotypes &times; positions.

If we wish to perform QTL calculations at positions between markers (so called "pseudomarkers"), we first need to insert such positions into the genetic map with the function `insert_pseudomarkers()`. Unlike [R/qtl], the map is kept separate from the genotype
probabilities.

We'll use the
[iron dataset](https://github.com/kbroman/qtl2/tree/gh-pages/assets/sampledata/iron)
from
[Grant et al. (2006) Hepatology 44:174-185](https://www.ncbi.nlm.nih.gov/pubmed/16799992)
(an intercross) as an example. We first load the data:

```{r load_data}
library(qtl2geno)
iron <- read_cross2( system.file("extdata", "iron.zip", package="qtl2geno") )
```

(_Note_: you can use `library(qtl2)` to load the
three main packages,
[qtl2geno](https://github.com/rqtl/qtl2geno),
[qtl2scan](https://github.com/rqtl/qtl2scan), and
[qtl2plot](https://github.com/rqtl/qtl2plot), all at once.)

We then use `insert_pseudomarkers()` to insert pseudomarkers into the
genetic map, which we grab from the `iron` object as `iron$gmap`:

```{r insert_pseudomarkers}
map <- insert_pseudomarkers(iron$gmap, step=1)
```

And next we use `calc_genoprob()` to calculate the QTL genotype probabilities.

```{r calc_genoprob}
pr <- calc_genoprob(iron, map, err=0.002)
```

To speed up the calculations with large datasets on a multi-core machine, you can use the argument `cores`. With `cores=0`, the number of available cores will be detected via `parallel::detectCores()`. Otherwise, specify the number of cores as a positive integer.

```{r calc_genoprob_multicore, eval=FALSE}
pr <- calc_genoprob(iron, map, err=0.002, cores=4)
```

The genome scan functions use genotype probabilities as well as a matrix of phenotypes. If you wished to perform a genome scan via an additive allele model, you would first convert the genotype probabilities to allele probabilities, using the function `genoprob_to_alleleprob()`.

```{r allele_probs}
apr <- genoprob_to_alleleprob(pr)
```
