---
title: "Calculating Genotype Probabilities"
teaching: 0
exercises: 0
questions:
- "How do I calculate QTL at positions between genotyped markers?"
- "How do I calculate QTL genotype probabilities?"
- "How do I calculate allele probabilities?"
- "How can I speed up calculations if I have a large data set?"
objectives:
- To explain why the first step in QTL analysis is to calculate genotype probabilities.
- To insert pseudomarkers between genotyped markers.
- To calculate genotype probabilities.
keypoints:
- "The first step in QTL analysis is to calculate genotype probabilities."
- "Insert pseudomarkers to calculate QTL at positions between genotyped markers."
- "Calculate genotype probabilities between genotyped markers with calc_genoprob()."
source: Rmd
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("06-")
```

The first basic task in QTL analysis is to calculate conditional genotype probabilities, given the observed marker data, at each putative QTL position. For example, the first step would be to determine the probabilities for genotypes AA and AB at the locus indicated below.

![](../fig/unknown_genotype.png)

Hidden Markov models (HMM) are used to calculate
QTL genotype probabilities, to simulate from the joint genotype distribution and to calculate the most likely sequence of underlying genotypes (all conditional on the observed marker data). This is done in a quite general way, with possible allowance for the presence of genotyping errors. For convenience we assume no crossover interference.

The `calc_genoprob()` function in the [qtl2geno](https://github.com/rqtl/qtl2geno)
package calculates QTL genotype probabilities, conditional on the available marker data. These are needed for most of the QTL mapping functions. Unlike the corresponding function in
[R/qtl](http://rqtl.org), `calc.genoprob()`, the result is not inserted back into the input cross object, but is returned as a list of three-dimensional arrays (one per chromosome). Each 3d array of probabilities is arranged as individuals &times; genotypes &times; positions.

[image of 3d array w/probs, inds, etc]

If we wish to perform QTL calculations at positions between markers (so called "pseudomarkers"), we first need to insert such positions into the genetic map with the function `insert_pseudomarkers()`. Unlike [R/qtl], the map is kept separate from the genotype
probabilities.

We'll use the
[iron dataset](https://github.com/kbroman/qtl2/tree/gh-pages/assets/sampledata/iron)
from
[Grant et al. (2006) Hepatology 44:174-185](https://www.ncbi.nlm.nih.gov/pubmed/16799992)
(an intercross) as an example. We first load the data:

```{r load_data}
library(qtl2geno)
iron <- read_cross2( system.file("extdata", "iron.zip", package="qtl2geno") )
```

(_Note_: you can use `library(qtl2)` to load the
three main packages,
[qtl2geno](https://github.com/rqtl/qtl2geno),
[qtl2scan](https://github.com/rqtl/qtl2scan), and
[qtl2plot](https://github.com/rqtl/qtl2plot), all at once.)

```{r summary_data}
summary(iron)
names(iron)
```

Have a look at the markers listed in the genetic map, `gmap`. Markers are listed by chromosome and described by cM position.

```{r map_data}
iron$gmap
```

We then use `insert_pseudomarkers()` to insert pseudomarkers into the
genetic map, which we grab from the `iron` object as `iron$gmap`:

```{r insert_pseudomarkers}
map <- insert_pseudomarkers(map=iron$gmap, step=1)
```

Now have a look at the new `map`.

```{r view_map}
map
```

Notice that pseudomarkers are now spaced at 1 cM intervals from genotyped markers. The argument `step=1` generated pseudomarkers at these intervals. 

Next we use `calc_genoprob()` to calculate the QTL genotype probabilities.

```{r calc_genoprob}
pr <- calc_genoprob(cross=iron, map=map, err=0.002)
```

View the first several rows of genotype probabilities for pseudomarker c1.loc28 on chromosome 1.

```{r view_genoprob}
head((pr$`1`)[,,"c1.loc28"])
```

**A bit more advanced (optional)** To speed up the calculations with large datasets on a multi-core machine, you can use the argument `cores`. With `cores=0`, the number of available cores will be detected via `parallel::detectCores()`. Otherwise, specify the number of cores as a positive integer.

```{r calc_genoprob_multicore, eval=FALSE}
pr <- calc_genoprob(cross=iron, map=map, err=0.002, cores=4)
```

**(optional)** The genome scan functions use genotype probabilities as well as a matrix of phenotypes. If you wished to perform a genome scan via an additive allele model, you would first convert the genotype probabilities to allele probabilities, using the function `genoprob_to_alleleprob()`.

```{r allele_probs}
apr <- genoprob_to_alleleprob(probs=pr)
```


> ## Challenge 1
> Explain why calculating genotype probabilities is the first step in QTL analysis.
>
> > ## Solution to Challenge 1
> >
> {: .solution}
{: .challenge}


> ## Challenge 2
> Load the grav2.zip file into an object called `grav`. 
> Insert pseudomarkers and calculate genotype probabilities.
>
> > ## Solution to Challenge 2
> >
> > grav <- read_cross2( system.file("extdata", "grav2.zip", package="qtl2geno") )
> > summary(grav)
> > grav$gmap
> > gravmap <- insert_pseudomarkers(map = grav$gmap, step=2)
> > gravpr <- calc_genoprob(grav, map = gravmap)
> > head((gravpr$`5`)[,,"c5.loc4"])
> {: .solution}
{: .challenge}



