---
title: "Performing a genome scan with a linear mixed model"
teaching: 20
exercises: 10
questions:
- "How do I use a linear mixed model in a genome scan?"
- "How do different mapping and kinship calculation methods differ?"
objectives:
- Create a genome scan with a linear mixed model.
- Compare LOD plots for Haley-Knott regression and linear mixed model methods.
- Compare LOD plots for the standard kinship matrix with the leave-one-chromosome-out (LOCO) method.
keypoints:
- "To perform a genome scan with a linear mixed model, supply a kinship matrix."
- "Different mapping and kinship calculation methods give different results."
source: Rmd
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("08-")
```

Genetic mapping in mice presents a good example of why accounting for population structure is important. Laboratory mouse strains are descended from a small number of founders (fancy mice) and went through several population bottlenecks. Wild-derived strains are not descended from fancy mice and don't share the same history as laboratory strains. Linear mixed models were developed to solve problems with population structure created by differing ancestries, and to handle relatedness between individuals.  Linear mixed models (LMMs) consider genome-wide similarity between all pairs of individuals to account for population structure, known kinship and unknown relatedness. Linear mixed models in mapping studies can successfully correct for genetic relatedness between individuals in a population by incorporating kinship into the model. Earlier we [calculated a kinship matrix](https://smcclatchy.github.io/mapping/04-calc-kinship/) for input to a linear mixed model to account for relationships among individuals. 

Recall that a linear model can take the form 

$$y = B_0 + B_iX + \epsilon$$ 
where $y$ is the response variable (phenotype in our case), $B_0$ is the y-intercept (mean phenotype), $X$ is the genotype, $B_i$ is a regression coefficient that in this case represents the slope of the line, and $\epsilon$ is the error. In the figure below, the linear model is $y = B_0 + B_iX + \epsilon$.

![](../fig/residual.png)

To perform a genome scan using a linear mixed model you also use the function `scan1`; you just need to provide the argument `kinship`, a kinship matrix (or, for the LOCO method, a list of kinship matrices).

```{r scan1_pg, eval=FALSE}
out_pg <- scan1(pr, iron$pheno, kinship=kinship, Xcovar=Xcovar)
```

Again, on a multi-core machine, you can get some speed-up using the `cores` argument.

```{r scan1_pg_multicore, eval=FALSE}
out_pg <- scan1(pr, iron$pheno, kinship, Xcovar=Xcovar, cores=4)
```

For the LOCO (leave one chromosome out) method, provide the list of kinship matrices as obtained from `calc_kinship()` with `method="loco"`.

```{r scan1_pg_loco, eval=FALSE}
out_pg_loco <- scan1(pr, iron$pheno, kinship_loco, Xcovar=Xcovar)
```

To plot the results, we again use `plot_scan1()` from the [qtl2plot](https://github.com/rqtl/qtl2plot) package, or just type `plot()`.

Here is a plot of the LOD scores, by Haley-Knott regression and the linear mixed model using either the standard kinship matrix or the LOCO method.

```{r plot_lod_scan1_vs_pg, eval=FALSE}
color <- c("slateblue", "violetred", "green3")
par(mar=c(4.1, 4.1, 1.6, 1.1))
ymx <- max(maxlod(out), maxlod(out_pg), maxlod(out_pg_loco))
for(i in 1:2) {
    plot(out, map, lodcolumn=i, col=color[1], main=colnames(iron$pheno)[i],
              ylim=c(0, ymx*1.02))
    plot(out_pg, map, lodcolumn=i, col=color[2], add=TRUE)
    plot(out_pg_loco, map, lodcolumn=i, col=color[3], add=TRUE, lty=2)
    legend("topleft", lwd=2, col=color, c("H-K", "LMM", "LOCO"), bg="gray90", lty=c(1,1,2))
}
```

For the liver phenotype (top panel), the three methods give quite different results. The linear mixed model with an overall kinship matrix gives much lower LOD scores than the other two methods.  On chromosomes with some evidence of a QTL, the LOCO method gives higher LOD scores than Haley-Knott, except on chromosome 16 where it gives lower LOD scores.

For the spleen phenotype (bottom panel), the linear mixed model with an overall kinship matrix again gives much lower LOD scores than the other two methods. However, in this case Haley-Knott regression and the LOCO method give quite similar results.
