---
title: "SNP association"
teaching: 0
exercises: 0
questions:
- "?"
objectives:
- 
- 
keypoints:
- "."
- "."
source: Rmd
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("11-")
```

## SNP association

For multi-parent crosses, it can be useful to collapse the genotype or
allele probabilities according to the founder genotypes of the various
SNPs in the region of a QTL.

### QTL analysis in Diversity Outbred mice

To illustrate this sort of SNP association analysis, we'll consider some
Diversity Outbred mouse data. The Diversity Outcross (DO) mice are an
advanced intercross population derived from the same eight founder strains
as the Collaborative Cross (CC). See
[Svenson et al. (2012)](https://www.ncbi.nlm.nih.gov/pubmed/22345611)
and
[Gatti et al. (2014)](https://www.ncbi.nlm.nih.gov/pubmed/25237114).

We'll consider a subset of the data from
[Recla et al. (2014)](https://www.ncbi.nlm.nih.gov/pubmed/24700285),
available as part of the
[qtl2data github repository](https://github.com/rqtl/qtl2data). (The
full data are in
[`DO_Recla`](https://github.com/rqtl/qtl2data/tree/master/DO_Recla);
the directory
[`DOex`](https://github.com/rqtl/qtl2data/tree/master/DOex) contains a
reduced set, with just three chromosomes, one phenotype
(`OF_immobile_pct`, percent immobile in the open field test), and a
reduced set of markers.

You can download the data from a single zip file, as follows:

```{r read_DOex_data}
file <- paste0("https://raw.githubusercontent.com/rqtl/",
               "qtl2data/master/DOex/DOex.zip")
DOex <- read_cross2(file)
```

Let's quickly whip through a basic analysis.

We first calculate genotype probabilities and convert them to allele
probabilities. We'll just use marker locations and not insert any
pseudomarkers.

```{r DOex_calc_genoprob, eval=FALSE}
pr <- calc_genoprob(DOex, error_prob=0.002)
apr <- genoprob_to_alleleprob(pr)
```

```{r download_alleleprobs, echo=FALSE, results="hide"}
tmpfile <- tempfile()
file <- paste0("https://raw.githubusercontent.com/rqtl/",
               "qtl2data/master/DOex/DOex_alleleprobs.rds")
download.file(file, tmpfile, quiet=TRUE)
apr <- readRDS(tmpfile)
unlink(tmpfile)
```

We calculate kinship matrices (using the "loco" method, though with the
caveat that here we are only considering genotypes on three chromosomes).

```{r DOex_calc_kinship}
k <- calc_kinship(apr, "loco")
```

We create a numeric covariate for sex; be sure to include the individual
IDs as names.

```{r DOex_create_sex_covar}
sex <- (DOex$covar$Sex == "male")*1
names(sex) <- rownames(DOex$covar)
```

We perform a genome scan with a linear mixed model (adjusting for
a residual polygenic effect), with sex as an additive covariate.

```{r DOex_scan1_pg}
out <- scan1(apr, DOex$pheno, k, sex)
```

Here's a plot of the results.

```{r plot_DOex_scan}
par(mar=c(4.1, 4.1, 0.6, 0.6))
plot(out, DOex$gmap)
```

There's a strong peak on chromosome 2. Let's look at the QTL effects.
We estimate them with `scan1coef()`. We need to subset the allele
probabilities and the list of kinship matrices.

```{r DOex_effects_c2}
coef_c2 <- scan1coef(apr[,"2"], DOex$pheno, k[["2"]], sex)
```

For the DO, with 8 QTL alleles, we can use the function `plot_coefCC`
in the [R/qtl2plot](https://github.com/rqtl/qtl2plot) package, which
plots the 8 allele effects in the "official" Collaborative Cross (CC)
colors. (Well, actually _slightly_ modified colors, because I think the
official colors are kind of ugly.) The strong locus seems to be mostly
due to the NZO allele. Note that `CCcolors` is a vector of colors
included in the qtl2plot package; there's also a `CCorigcolors` object
with the _official_ colors.

```{r plot_DOex_effects}
par(mar=c(4.1, 4.1, 0.6, 0.6))
plot_coefCC(coef_c2, DOex$gmap["2"], bgcolor="gray95")
legend("bottomleft", col=CCcolors, names(CCcolors), ncol=2, lwd=2, bg="gray95")
```

### SNP associations

Okay, now finally we get to the SNP associations. We have a large peak
on chromosome 2, and we want to look at individual SNPs in the region
of the locus.

Well, actually, we first need to find the location of the inferred
QTL.  The peak LOD score on chromosome 2 occurs at
`r round(max(out, DOex$gmap, chr="2")$pos, 1)` cM. But to find nearby SNPs, we really want
to know the Mbp position. The calculations were only performed at the
marker positions, and so we need to find the peak marker and then
find it's physical location:

```{r DOex_find_peak_in_Mbp}
marker <- rownames(max(out, DOex$gmap, chr="2"))
peak_Mbp <- DOex$pmap[["2"]][marker]
```

The marker is at `r round(peak_Mbp, 1)` Mbp.

Now we need to identify the SNPs in this region. We'll focus on a 2
Mbp interval centered at `r round(peak_Mbp, 1)` Mbp. We're still
working on how best to quickly access SNP data. In the meantime, we
can grab a predefined table of SNPs that's available in the
[qtl2data repository](https://github.com/rqtl/qtl2data). It's saved as
an RDS file, which is a slight hassle to load over the web.

```{r load_c2_snps}
tmpfile <- tempfile()
file <- "https://raw.githubusercontent.com/rqtl/qtl2data/master/DOex/c2_snpinfo.rds"
download.file(file, tmpfile, quiet=TRUE)
snpinfo <- readRDS(tmpfile)
unlink(tmpfile)
```

Here's the first few rows of the data. The columns are the SNP name,
the chromosome, the Mbp position (in Mouse genome build 38), the
alleles (with the B6 allele before the `|` and any other alleles
after; in the case of multiple alternate alleles, they are separated
by `/`). Finally, there are eight columns of genotypes for the 8 CC
founder strains, coded as `1`/`3`.

```{r head_snpinfo}
head(snpinfo)
```

We first convert the founder genotypes to a "strain distribution
pattern" (SDP): an integer whose binary encoding corresponds to the 8
founders' genotypes.

```{r calc_sdp}
snpinfo$sdp <- calc_sdp(snpinfo[,-(1:4)])
```

We've added the SDP as an additional column.

```{r head_snpinfo_with_sdp}
head(snpinfo)
```

(Note that there's also a function `invert_sdp()` for converting the
SDPs back into founder genotypes.)

To perform the SNP association analysis, we first use the allele
probabilities and the founder SNP genotypes to infer the SNP genotypes
for the DO mice. That is, at each SNP, we want to collapse the eight
founder allele probabilities to two SNP allele probabilities, using
the the SNP genotypes of the founders.

We do this assuming that the allele probabilities were
calculated sufficiently densely that they can be assumed to be
constant in intervals. With this assumption, we then:

- Find the interval for each SNP.
- Reduce the SNPs to a "distinct" set: if two SNPs have the same SDP
  and are in the same interval, by our assumption their allele
  probabilities will be the same.
- Take the average of the allele probabilities at the two endpoints of
  each interval.
- Collapse the 8 allele probabilities to two according to each
  observed SDP in the interval.

We further create a look-up table relating the full set of SNPs to the
reduced set (one of each observed SDP in each interval).

We first need to identify the equivalent SNPs, using the function
`index_snps()`. This requires a physical map of the
markers/pseudomarkers used to calculate the genotype probabilities.
We take this directly from the `DOex`
object, as we'd calculated the allele
probabilities only at the observed markers. If we'd also calculated
probabilities at pseudomarker positions between markers, we'd need to
use interpolation to get Mbp positions for the
pseudomarkers. There's a function `interp_map()` for assisting with
that.

The `index_snps()` function takes the physical map and the `snpinfo`
data frame, include the strain distribution patterns we calculated above.
It inserts three new columns into the data frame (`"index"`,
`"interval"`, and `"on_map"`: indexes to a set of non-equivalent SNPs,
map intervals in which the SNPs lie, and whether the SNPs correspond
to marker/pseudomarker positions).

```{r index_snps}
snpinfo <- index_snps(DOex$pmap, snpinfo)
```

We can then use the function `genoprob_to_snpprob()`,
which takes the allele probabilities (or the full genotype
probabilities, if you want to use a full 3-genotype model at each
SNP), to collapse the genotype probabilities to SNP genotype
probabilities.

```{r calc_snpprobs}
snp_pr <- genoprob_to_snpprob(apr, snpinfo)
```

The output of this function, `snp_pr`, has the same form as the input
`apr` object with allele probabilities, and can be used directly in a
call to `scan1()`. And so we can now use the object to perform the SNP
association analysis in the region, using the same linear mixed model.
We need to be sure to use the correct kinship matrix.

```{r perform_snp_scan}
out_snps <- scan1(snp_pr, DOex$pheno, k[["2"]], sex)
```

The function `plot_snpasso()` in the qtl2plot package can be used to
plot the results, with points at each of the SNPs. The default is to
plot **all** SNPs: We calculated LOD scores only at a set of distinct
SNPs, but SNPs in the same interval with the same SDP will have the
same LOD score. It takes the `scan1()` output plus the `snpinfo` data
frame.

```{r plot_snp_asso}
par(mar=c(4.1, 4.1, 0.6, 0.6))
plot_snpasso(out_snps, snpinfo)
```

To get a table of the SNPs with the largest LOD scores, use the
function `top_snps()`. This will show all SNPs with LOD score within
some amount (the default is 1.5) of the maximum SNP LOD score.

```{r top_snps}
top_snps(out_snps, snpinfo)
```

The top SNPs all have NZO and CAST with a common allele, different
from the other 6 founders. The next-best SNPs have NZO with a unique
allele. Note that there's one SNP with two alternate alleles
(`C|G/T`). We are requiring that SNPs have just two alleles, and so we
group the alternate alleles together, though there's not a good reason
for this.

We can highlight these top SNPs in the SNP association plot using the
`drop` argument.

```{r plot_snp_asso_hilit}
par(mar=c(4.1, 4.1, 0.6, 0.6))
plot_snpasso(out_snps, snpinfo, drop=1.5)
```
